### PR TITLE
SNAT metrics

### LINKED ISSUE #35621: nat: CT creation and SNAT src port allocation failure metrics
For operations, it would be ideal to have metrics on the following error conditions:

* In the event of running [out of SNAT source ports](https://github.com/cilium/cilium/blob/e2b17ece53309060572db0eafffda14be96a97f8/bpf/lib/nat.h#L258-L262), the signal will likely trigger GC more often than usual.
* if new CT entries [can't be created](https://github.com/cilium/cilium/blob/e2b17ece53309060572db0eafffda14be96a97f8/bpf/lib/conntrack.h#L1068-L1071).

### PR DESCRIPTION
**Pull request by gentoo-root**
Collect a histogram of the number of attempts taken before a successful
(or an unsuccessful) port allocation for SNAT occurs. These data will be
useful for troubleshooting and for evaluating SNAT port allocation
algorithms. For example, collecting these statistics at various ratios
of allocated ports will show the distribution of how quickly the
algorithm can find an unused port (and whether at all).

New cilium-dbg commands are added:

`cilium-dbg bpf nat retries list` - to print the histogram.
`cilium-dbg bpf nat retries flush` - to reset the counters to zero.

Fixes: #35621

```release-note
Collect a histogram and show the distribution of the number of attempts needed to allocate a port for SNAT.
```

### CODE REVIEW

File: cilium-dbg/cmd/bpf_nat_retries_list.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+)
+
+// bpfNatRetriesListCmd represents the bpf_nat_retries_list command
+var bpfNatRetriesListCmd = &cobra.Command{
+	Use:     "list",
+	Aliases: []string{"ls"},
+	Short:   "Show the NAT retries histogram",
+	Run: func(cmd *cobra.Command, args []string) {
+		common.RequireRootPrivilege("cilium bpf nat retries list")
+		if len(args) == 0 {
```
Comment (derailed) [inline cilium-dbg/cmd/bpf_nat_retries_list.go]: nit: Better to exit early if extra invalid args are detected

File: cilium-dbg/cmd/bpf_nat_retries_list.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	},
+}
+
+func init() {
+	BPFNatRetriesCmd.AddCommand(bpfNatRetriesListCmd)
+	command.AddOutputOption(bpfNatRetriesListCmd)
+}
+
+func dumpRetries(m nat.RetriesMap) {
+	entries := make([]nat.RetriesMapRecord, 0)
```
Comment (derailed) [inline cilium-dbg/cmd/bpf_nat_retries_list.go]: nit: move where it's actually used. Also nil slice would suffice.

File: pkg/bpf/map_linux.go [hunk 3]
```diff
@@ -54,6 +54,12 @@ type MapValue interface {
 	New() MapValue
 }
 
+type MapPerCPUValue interface {
```
Comment (derailed) [inline pkg/bpf/map_linux.go:59]: nit: docs

File: pkg/bpf/map_linux.go [hunk 4]
```diff
@@ -633,6 +644,35 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
 	return i.Err()
 }
 
+type DumpPerCPUCallback func(key MapKey, values any)
```
Comment (derailed) [inline pkg/bpf/map_linux.go:651]: nit: docs

File: pkg/bpf/map_linux.go [hunk 5]
```diff
@@ -633,6 +644,35 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
...
+
+func (m *Map) DumpPerCPUWithCallback(cb DumpPerCPUCallback) error {
```
Comment (derailed) [inline pkg/bpf/map_linux.go:657]: nit: docs

File: pkg/bpf/map_linux.go [hunk 6]
```diff
@@ -633,6 +644,35 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
...
+	if cb == nil {
+		return errors.New("empty callback")
+	}
+
+	if err := m.Open(); err != nil {
```
Comment (derailed) [inline pkg/bpf/map_linux.go:671]: Not sure of implications here but the map is open and never closed and the path is not checked? 
It looks like the call site also manages existence/open/close on the map?
Do we want to assume call site is responsible for these checks if existence is required?
Would it be best to let this method handle it instead and perhaps DumpPerCPUWithCBIfExists?

File: pkg/maps/nat/nat.go [hunk 7]
```diff
@@ -371,3 +420,16 @@ func maxEntries() int {
 	}
 	return option.LimitTableMax
 }
+
+func RetriesMaps(ipv4, ipv6, nodeport bool) (ipv4RetriesMap, ipv6RetriesMap RetriesMap) {
```
Comment (derailed) [inline pkg/maps/nat/nat.go:427]: nit: docs

**Review by derailed — COMMENTED** [review]
@gentoo-root Nice work! Could be off base here so please disregard if not applicable...

File: cilium-dbg/cmd/bpf_nat_retries_list.go (hunk 1 above)
Comment (gentoo-root) [reply cilium-dbg/cmd/bpf_nat_retries_list.go]: Inspired by [this](https://github.com/cilium/cilium/blob/181ed5a73517ba9e3ca0c64063202049d6f32321/cilium-dbg/cmd/bpf_nat_list.go#L28-L52), but I agree with you.

File: cilium-dbg/cmd/bpf_nat_retries_list.go (hunk 2 above)
Comment (gentoo-root) [reply cilium-dbg/cmd/bpf_nat_retries_list.go]: ACK (inspired by [this](https://github.com/cilium/cilium/blob/181ed5a73517ba9e3ca0c64063202049d6f32321/cilium-dbg/cmd/bpf_nat_list.go#L62))

File: pkg/bpf/map_linux.go (hunk 6 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux.go:671]: > the map is open and never closed

It's similar to other methods ([DumpWithCallback](https://github.com/cilium/cilium/blob/181ed5a73517ba9e3ca0c64063202049d6f32321/pkg/bpf/map_linux.go#L614), [Lookup](https://github.com/cilium/cilium/blob/181ed5a73517ba9e3ca0c64063202049d6f32321/pkg/bpf/map_linux.go#L1028)). The map is opened on demand on the first access, and it's the responsibility of the caller to close it. It's the pattern that is widely used across the codebase, and changing it would require rewriting of a lot of things. I'm following the existing pattern. E.g., like in `bpf nat list`: [link](https://github.com/cilium/cilium/blob/181ed5a73517ba9e3ca0c64063202049d6f32321/cilium-dbg/cmd/bpf_nat_list.go#L68-L89). I do it in a similar fashion in `bpf nat retries list`: [link](https://github.com/cilium/cilium/pull/36730/files/0b36a35aa4b7b661ee6070e4e13bbba3cbf293be#diff-8c02757c3e3dd9daf80a1155d6e8947d7167dace77e8526253a827a41d713278R48).

> It looks like the call site also manages existence/open/close on the map?
Do we want to assume call site is responsible for these checks if existence is required?

Basically, yes, and it's an assumption that we make in the existing code, I'm not adding anything new here.

**Review by borkmann — APPROVED** [review]
apart from comments from Fernand, this lgtm! Maybe also add a commit adding `cilium-dbg bpf nat retries list` to the sysdump?

File: pkg/bpf/map_linux.go (hunk 3 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux.go:59]: ACK.

File: pkg/bpf/map_linux.go (hunk 4 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux.go:651]: ACK.

File: pkg/bpf/map_linux.go (hunk 5 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux.go:657]: ACK.

File: pkg/maps/nat/nat.go (hunk 7 above)
Comment (gentoo-root) [reply pkg/maps/nat/nat.go:427]: ACK.

File: pkg/bpf/map_linux.go (hunk 3 above)
Comment (derailed) [reply pkg/bpf/map_linux.go:59]: Exported calls docs should always start with the item name i.e MapPerCPUValue.

File: pkg/bpf/map_linux.go (hunk 4 above)
Comment (derailed) [reply pkg/bpf/map_linux.go:651]: Same as above ^

File: pkg/bpf/map_linux.go (hunk 6 above)
Comment (derailed) [reply pkg/bpf/map_linux.go:671]: @gentoo-root Thank you for the clarification! Are there implications in opening the map twice? The current pattern seems off as the call site opens the resource and is responsible to close it but the underlying call re-open the resource and assumes the call site is responsible for closing it?? 
This might be ok on a short lived call but it seems to be an anti-pattern.

**Review by derailed — COMMENTED** [review]
@gentoo-root Thanks for the updates!

File: pkg/bpf/map_linux.go [hunk 8]
```diff
@@ -1272,6 +1319,46 @@ func (m *Map) DeleteAll() error {
...
+		if metrics.BPFMapOps.IsEnabled() {
+			metrics.BPFMapOps.WithLabelValues(m.commonName(), metricOpUpdate, metrics.Error2Outcome(err)).Inc()
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	err := i.Err()
```
Comment (tommyp1ckles) [inline pkg/bpf/map_linux.go]: nit: can just be `return i.Err()`

File: pkg/maps/nat/nat.go [hunk 9]
```diff
@@ -71,6 +79,14 @@ type NatMap interface {
 	DumpWithCallback(bpf.DumpCallback) error
 }
 
+type RetriesMap interface {
+	Open() error
+	Close() error
+	Path() (string, error)
+	DumpPerCPUWithCallback(bpf.DumpPerCPUCallback) error
```
Comment (tommyp1ckles) [inline pkg/maps/nat/nat.go:88]: minor/non-blocking: There's a bit of overlap with the `NatMap` interface, we could probably simplify using interface composition.

```go
type commonMap interface {
	Open() error
	Close() error
	Path() (string, error)
	DumpEntries() (string, error)  
}

type NatMap interface {
        commonMap
        DumpWithCallback(bpf.DumpCallback) error
}

type RetriesMap interface {
        commonMap
        DumpPerCPUWithCallback(bpf.DumpPerCPUCallback) error
        ClearAll() error
}
```

File: cilium-dbg/cmd/bpf_nat_retries_list.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		}
+	},
+}
+
+func init() {
+	BPFNatRetriesCmd.AddCommand(bpfNatRetriesListCmd)
+	command.AddOutputOption(bpfNatRetriesListCmd)
+}
+
+func dumpRetries(m nat.RetriesMap) {
```
Comment (tommyp1ckles) [inline cilium-dbg/cmd/bpf_nat_retries_list.go:45]: Question: Is there any plan to provide this as a prometheus metric, this would map well to proms histogram type.

**Review by tommyp1ckles — CHANGES_REQUESTED** [review]

**Review by tommyp1ckles — APPROVED** [review]
Just a few small points (should've added them as Comment not Require-Changes). Otherwise no blocking issues for me.

File: pkg/maps/nat/nat.go [hunk 11]
```diff
@@ -99,6 +115,39 @@ func NewMap(name string, family IPFamily, entries int) *Map {
...
+
+type RetriesValue struct {
+	Value uint32
+}
+
+type RetriesValues []RetriesValue
+
+func (k *RetriesValue) String() string    { return fmt.Sprintf("%d", k.Value) }
+func (k *RetriesValue) New() bpf.MapValue { return &RetriesValue{} }
+func (k *RetriesValue) NewSlice() any     { return &RetriesValues{} }
```
Comment (tommyp1ckles) [inline pkg/maps/nat/nat.go:135]: Actually, one question, shouldn't this return a pointer to a slice type?

**Review by tommyp1ckles — CHANGES_REQUESTED** [review]

File: pkg/bpf/map_linux.go [hunk 12]
```diff
@@ -633,6 +645,41 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
...
+
+// DumpPerCPUWithCallback iterates over the Map and calls the given
+// DumpPerCPUCallback for each map entry, passing the slice with all values
+// from all CPUs. With the current implementation, it is safe for callbacks
+// to retain the values received, as they are guaranteed to be new instances.
+func (m *Map) DumpPerCPUWithCallback(cb DumpPerCPUCallback) error {
+	if cb == nil {
+		return errors.New("empty callback")
+	}
+
```
Comment (rgo3) [inline pkg/bpf/map_linux.go:661]:
```suggestion

	// Validate map type early
	if !m.hasPerCPUValue() {
		return fmt.Errorf("map %s is not a per-CPU map", m.name)
	}

    // do type assertion once, avoid repetitive type assertions below in the loop
    // also bails out early if value type doesn't satisfy the MapPerCPUValue interface
	perCPUValue, ok := m.value.(MapPerCPUValue)
	if !ok {
		return fmt.Errorf("map %s value type does not implement MapPerCPUValue", m.name)
	}
```

**Review by rgo3 — CHANGES_REQUESTED** [review]
Should we add unit tests for the newly added `DumpPerCPUWithCallback` and `ClearAll` functions in the `bpf` package?

File: pkg/bpf/map_linux.go (hunk 6 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux.go:671]: > Are there implications in opening the map twice?

Open() is idempotent, the second call will be a no-op: https://github.com/cilium/cilium/blob/a9b6d692f62e56bff501ea9565fbb4bd91bab68a/pkg/bpf/map_linux.go#L547-L549

> it seems to be an anti-pattern

I tend to agree with you. If the open-close lifecycle is managed externally, the methods should fail when the map was not opened. If the methods try to fall back and open it, I see a bunch of problems: the caller doesn't know whether it needs to close it now (i.e. whether it was opened now or before), the method might use a wrong way of opening the map (there are Open, OpenMap, OpenOrCreate).

However, for some reason we have it in a lot of methods (DumpWithCallback, DumpReliablyWithCallback, Lookup, Update, delete, DeleteAll, resolveErrors), so it might be problematic to just get rid of it, because some other code might rely on this behavior. I also don't know the reason why we have it there in the first place. So, I'd rather keep the existing (anti-)pattern in this PR, and we can experiment with removing it in a follow-up (feel free to try it, as you are the one who spotted it).

File: pkg/maps/nat/nat.go (hunk 11 above)
Comment (gentoo-root) [reply pkg/maps/nat/nat.go:135]: You mean, instead of `any`? It's an interface method, and the caller needs a slice of structs, not a slice of interfaces. As the struct type can be anything that implements MapValue, and generic methods are not supported in Go, I had to erase the type. If you know a better way (without making [Map](https://github.com/cilium/cilium/blob/663c450daa6c6c80e9722637b5fa7390a64b9799/pkg/bpf/map_linux.go#L65) a generic struct and rewriting all code in the project), please tell me, I would be happy to change it into something better.

File: cilium-dbg/cmd/bpf_nat_retries_list.go (hunk 10 above)
Comment (gentoo-root) [reply cilium-dbg/cmd/bpf_nat_retries_list.go:45]: That's a good idea, and I guess it would be useful to expose it in prometheus. I'll check if it can be done here or as a follow-up. If you have an example of something similar, could you send me a pointer?

File: pkg/bpf/map_linux.go (hunk 12 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux.go:661]: Thanks! Makes sense to me, I'll add it.

File: pkg/maps/nat/nat.go (hunk 11 above)
Comment (tommyp1ckles) [reply pkg/maps/nat/nat.go:135]: Oh I misread, didn't realize RetriesValues was actually an alias for a slice type.

looks good

**Review by tommyp1ckles — APPROVED** [review]

File: pkg/bpf/map_linux.go [hunk 13]
```diff
@@ -1101,7 +1158,12 @@ func (m *Map) Update(key MapKey, value MapValue) error {
 		return err
 	}
 
-	err = m.m.Update(key, value, ebpf.UpdateAny)
+	if m.hasPerCPUValue() {
```
Comment (rgo3) [inline pkg/bpf/map_linux.go]: Not sure if we should make changes to the Update API just to facilitate a test.
Instead of using the Update API in the test, we can try to write a setup function that accesses the underlying `ebpf.Map` directly (I'll make a comment with my idea at the appropriate place in the code) inserting per-CPU values like in this test: https://github.com/cilium/ebpf/blob/f224bff41d4d89fca11ddac24453f6dacffb6f89/map_test.go#L2327. Once the map is setup check if Dump and Clear work as expected?

File: pkg/bpf/map_linux_test.go [hunk 14]
```diff
@@ -508,6 +536,51 @@ func TestDump(t *testing.T) {
...
 
+func TestDumpPerCPU(t *testing.T) {
+	testMap := setupPerCPU(t)
+
+	key1 := &TestKey{Key: 0}
+	value1 := &TestValue{Value: 205}
+	key2 := &TestKey{Key: 2}
+	value2 := &TestValue{Value: 206}
+
+	err := testMap.Update(key1, value1)
```
Comment (rgo3) [inline pkg/bpf/map_linux_test.go]: Would `testMap.m.Put(...)` work here?

**Review by rgo3 — CHANGES_REQUESTED** [review]
Made some suggestions to avoid changing the `Update` API to facilitate the unit tests.

File: pkg/bpf/map_linux.go (hunk 13 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux.go]: > write a setup function that accesses the underlying ebpf.Map directly

Good idea, thanks. Implemented.

File: pkg/bpf/map_linux_test.go [hunk 15]
```diff
@@ -508,6 +536,55 @@ func TestDump(t *testing.T) {
...
 
+func TestDumpPerCPU(t *testing.T) {
+	testMap := setupPerCPU(t)
+
+	key1 := &TestKey{Key: 0}
+	value1 := &TestValue{Value: 205}
+	key2 := &TestKey{Key: 2}
+	value2 := &TestValue{Value: 206}
+
+	func() {
```
Comment (rgo3) [inline pkg/bpf/map_linux_test.go:547]: Is there a benefit to wrapping this into an anonymous function?

**Review by rgo3 — APPROVED** [review]
LGTM, thanks!

File: pkg/bpf/map_linux_test.go (hunk 15 above)
Comment (gentoo-root) [reply pkg/bpf/map_linux_test.go:547]: It's for `defer testMap.lock.Unlock()` to trigger after all Update()s, but not later. As far as I know, it defers until the return from the function, so a regular `{ block of code }` wouldn't suffice.

File: pkg/bpf/map_linux.go (hunk 6 above)
Comment (derailed) [reply pkg/bpf/map_linux.go:671]: @gentoo-root Thanks for the clarifications! I think at the very least we should add docs to that effect on all the public APIs to ensure the call site is aware and does the right thing with that resource. 
And perhaps a TODO might be appropriate here so we can revisit for a code debt pass

**Review by derailed — APPROVED** [review]
@gentoo-root Tx for the updates!

File: bpf/node_config.h [hunk 16]
```diff
@@ -139,6 +139,10 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 #define CAPTURE6_SIZE 16384
 #endif /* ENABLE_IPV6 */
 
+#ifdef ENABLE_NODEPORT
+#define SNAT_COLLISION_RETRIES 128
+#endif
+
```
Comment (julianwiedmann) [inline bpf/node_config.h:145]: Looks like this change tripped over a weird issue in BPF program compilation during downgrade:
https://github.com/cilium/cilium/issues/37602

Comment (gentoo-root) [reply bpf/node_config.h:145]: The issue description says it happened on one node only, where node_config.h wasn't regenerated for some reason. Sounds like the general bug is with node_config.h generation, not directly caused by moving this #define.
